### PR TITLE
RE-36 Thaw adjustments

### PR DESCRIPTION
**Pull request by mattt416**
This commit makes some minor adjustments to assist with the newton thaw
process.

Issue: [RE-36](https://rpc-openstack.atlassian.net/browse/RE-36)

### CODE REVIEW

File: gating/thaw/restart-nova-api-metadata.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+- name: Bring nova-api-metadata back online
+  hosts: nova_api_metadata
+  tasks:
+    - name: Wait for nova-api-metadata service to time out
+      pause:
+        minutes: 2
```
Comment (prometheanfire) [inline gating/thaw/restart-nova-api-metadata.yml]: seems racy...  Is there a reason why it has to be this way?

Comment (mattt416) [reply gating/thaw/restart-nova-api-metadata.yml]: So when we build a deploy from a saved image, the containers are all down and therefore have to be started.  When the nova_api_metadata container starts, the nova-api-metadata service times out after a little bit of time with this error:

```
2018-02-07 21:02:55.165 196 DEBUG oslo_messaging._drivers.amqpdriver [req-151ddc90-b579-447e-a4dd-8d8aac7623ef - - - - -] CALL msg_id: 2e9721b51aed495397917916d0e655bd exchange 'nova' topic 'conductor' _send /openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_messaging/_drivers/amqpdriver.py:568
2018-02-07 21:03:55.168 196 ERROR oslo_service.service [req-151ddc90-b579-447e-a4dd-8d8aac7623ef - - - - -] Error starting thread.
2018-02-07 21:03:55.168 196 ERROR oslo_service.service Traceback (most recent call last):
2018-02-07 21:03:55.168 196 ERROR oslo_service.service   File "/openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_service/service.py", line 708, in run_service
2018-02-07 21:03:55.168 196 ERROR oslo_service.service     service.start()
2018-02-07 21:03:55.168 196 ERROR oslo_service.service   File "/openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/nova/service.py", line 357, in start
2018-02-07 21:03:55.168 196 ERROR oslo_service.service     self.binary)
2018-02-07 21:03:55.168 196 ERROR oslo_service.service   File "/openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_versionedobjects/base.py", line 177, in wrapper
2018-02-07 21:03:55.168 196 ERROR oslo_service.service     args, kwargs)
2018-02-07 21:03:55.168 196 ERROR oslo_service.service   File "/openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/nova/conductor/rpcapi.py", line 236, in object_class_action_versions
2018-02-07 21:03:55.168 196 ERROR oslo_service.service     args=args, kwargs=kwargs)
2018-02-07 21:03:55.168 196 ERROR oslo_service.service   File "/openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_messaging/rpc/client.py", line 169, in call
2018-02-07 21:03:55.168 196 ERROR oslo_service.service     retry=self.retry)
2018-02-07 21:03:55.168 196 ERROR oslo_service.service   File "/openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_messaging/transport.py", line 97, in _send
2018-02-07 21:03:55.168 196 ERROR oslo_service.service     timeout=timeout, retry=retry)
2018-02-07 21:03:55.168 196 ERROR oslo_service.service   File "/openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_messaging/_drivers/amqpdriver.py", line 584, in send
2018-02-07 21:03:55.168 196 ERROR oslo_service.service     retry=retry)
2018-02-07 21:03:55.168 196 ERROR oslo_service.service   File "/openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_messaging/_drivers/amqpdriver.py", line 573, in _send
2018-02-07 21:03:55.168 196 ERROR oslo_service.service     result = self._waiter.wait(msg_id, timeout)
2018-02-07 21:03:55.168 196 ERROR oslo_service.service   File "/openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_messaging/_drivers/amqpdriver.py", line 456, in wait
2018-02-07 21:03:55.168 196 ERROR oslo_service.service     message = self.waiters.get(msg_id, timeout=timeout)
2018-02-07 21:03:55.168 196 ERROR oslo_service.service   File "/openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_messaging/_drivers/amqpdriver.py", line 347, in get
2018-02-07 21:03:55.168 196 ERROR oslo_service.service     'to message ID %s' % msg_id)
2018-02-07 21:03:55.168 196 ERROR oslo_service.service MessagingTimeout: Timed out waiting for a reply to message ID 2e9721b51aed495397917916d0e655bd
2018-02-07 21:03:55.168 196 ERROR oslo_service.service
2018-02-07 21:03:55.171 196 INFO nova.wsgi [-] Stopping WSGI server.
2018-02-07 21:03:55.171 196 DEBUG oslo_concurrency.lockutils [-] Acquired semaphore "singleton_lock" lock /openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_concurrency/lockutils.py:212
2018-02-07 21:03:55.172 196 DEBUG oslo_concurrency.lockutils [-] Releasing semaphore "singleton_lock" lock /openstack/venvs/nova-r14.7.0/lib/python2.7/site-packages/oslo_concurrency/lockutils.py:225
```
If you try to restart it manually before it has timed out, you basically just get that same error. What seems to work is to just give it some time to error itself, and then restart it again.  I know it's not the most elegant solution but I'm not entirely sure why it's erroring like that.  I've tried bringing the containers in different order (rabbit first, for example) but nothing seemed to help.

On a side note, when it times out, the process exits cleanly, so systemd doesn't try to restart it.

Comment (prometheanfire) [reply gating/thaw/restart-nova-api-metadata.yml]: sure, but is there no way other than a sleep?  what happens when we run this on a slower cloud?

Comment (odyssey4me) [reply gating/thaw/restart-nova-api-metadata.yml]: Not sure if this is why 2 mins makes, sense, but the timeout value for all the services are 2 mins: https://github.com/openstack/openstack-ansible-os_nova/blob/master/templates/nova-systemd-init.j2#L23

Comment (prometheanfire) [reply gating/thaw/restart-nova-api-metadata.yml]: ok, if the value were a var I'd say we should use that, but since it's hard coded this is fine

Comment (odyssey4me) [reply gating/thaw/restart-nova-api-metadata.yml]: If this startup issue is something not unique to this situation (imaging the host, etc) but instead a standard issue thing on any reboot, then it'd be good to get some sort of fix implemented upstream. I unfortunately can't think of anything off-hand though.

**Comment by byronmccollum** [comment]
Seeing the same issue with a simple `systemctl [stop|start] lxc`, or rebooting the host.

The problem appears to be that conductor doesn't get started before metadata API.

http://paste.openstack.org/show/670659/

If you manually start `galera, memcache, rabbit, keystone, nova_conductor` first, then metadata API comes up cleanly.

http://paste.openstack.org/show/670677/

**Comment by byronmccollum** [comment]
Startup order, for reference:

```
root@osa-aio:/var/lib/lxc# grep "lxc.start.order" */config | sort -k 3 -n
aio1_galera_container-922ff1be/config:lxc.start.order = 10
aio1_keystone_container-d879ec9b/config:lxc.start.order = 19
aio1_rabbit_mq_container-8971ece6/config:lxc.start.order = 19
aio1_rsyslog_container-28859039/config:lxc.start.order = 19
aio1_neutron_agents_container-b3cb3df6/config:lxc.start.order = 29
aio1_nova_api_metadata_container-2eba389b/config:lxc.start.order = 39
aio1_nova_api_os_compute_container-c2e5adf4/config:lxc.start.order = 39
aio1_nova_cert_container-57bf0d80/config:lxc.start.order = 39
aio1_nova_conductor_container-818581c7/config:lxc.start.order = 39
aio1_nova_console_container-607f0ccd/config:lxc.start.order = 39
aio1_nova_scheduler_container-d3b3c6ef/config:lxc.start.order = 39
aio1_swift_proxy_container-60d3ac51/config:lxc.start.order = 39
aio1_cinder_api_container-0fd9ee34/config:lxc.start.order = 100
aio1_cinder_scheduler_container-aae8da17/config:lxc.start.order = 100
aio1_glance_container-c00bda7f/config:lxc.start.order = 100
aio1_heat_apis_container-90195141/config:lxc.start.order = 100
aio1_heat_engine_container-3f9ed4da/config:lxc.start.order = 100
aio1_horizon_container-34d1dc6b/config:lxc.start.order = 100
aio1_memcached_container-715dc71b/config:lxc.start.order = 100
aio1_neutron_server_container-2c52c344/config:lxc.start.order = 100
aio1_repo_container-7c8cb43a/config:lxc.start.order = 100
aio1_utility_container-1d4ef857/config:lxc.start.order = 100
```

**Comment by mattt416** [comment]
@byronmccollum I tried modifying https://github.com/rcbops/rpc-openstack/blob/newton/gating/thaw/thaw.yml to start galera/rabbit first, then nova_conductor, and then nova_api_metadata, but it didn't work.  I suspect this is because we're just using lxc-start which ignores the start delay in the config file.  Your comment here actually raises a good point in that we're completely ignoring the start order and should resort to starting with lxc-autostart (once the containers can actually be started).

I'm testing this upstream change atm:

```
diff --git a/playbooks/common-playbooks/nova.yml b/playbooks/common-playbooks/nova.yml
index bee2b00..91233f4 100644
--- a/playbooks/common-playbooks/nova.yml
+++ b/playbooks/common-playbooks/nova.yml
@@ -51,6 +51,16 @@
       vars:
         extra_container_config_no_restart:
           - "lxc.start.order=39"
+      when:
+        - "'nova_conductor' in group_names"
+
+    - name: Configure container
+      include: ../common-tasks/os-lxc-container-setup.yml
+      vars:
+        extra_container_config_no_restart:
+          - "lxc.start.order=49"
+      when:
+        - "'nova_conductor' not in group_names"

     - name: Configure log directories (on metal)
       include: ../common-tasks/os-log-dir-setup.yml
```

I'm not sure if changing the start order for most nova containers from 39 to 49 is a problem, but I'm sure @odyssey4me will know if that's going to be a problem (assuming it even works).

**Comment by odyssey4me** [comment]
@mattt416 @byronmccollum Yeah, the start order for LXC needs adjusting to try and match the start order specified in https://github.com/openstack/openstack-ansible-os_nova/blob/master/defaults/main.yml#L435-L507 or something close to it. This is a really good find and something that's been a loose end in my mind for some time. I expect that the LXC start order doesn't have to be quite as detailed as some services will be more tolerant than others, but this is definitely something we should endeavour to resolve upstream.

**Comment by mattt416** [comment]
My patch above seems to work, so I've created https://review.openstack.org/#/c/544038/ to make this change upstream.

**Comment by mattt416** [comment]
I've removed the restart of nova-api-metadata from this patch.  There is a SHA bump in https://github.com/rcbops/rpc-openstack/pull/2854 which will pull in the container ordering fix in OSA, but I'll still need to create a separate patch to update the thaw process (for pike and newton) to start containers with lxc-autostart rather than manually with lxc-start. This PR should be fine to merge since it's a standalone fix that is specific to newton.

**Pull request merged**